core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/CockroachActionTests.java

```diff
@@ -0,0 +1,733 @@
+                // The status is typically NOTIFIED, but it can be "DONE" as well if a node didn't process the latest
+                // cluster state update yet
+                assertThat(status.getState().toString(), anyOf(equalTo("NOTIFIED"), equalTo("DONE")));
+            } else {
+                fail("Unexpected action " + taskInfo.getAction());
+            }
+        }
+
+        acknowledgementLatch.countDown();
+        assertTrue(finishResponseLatch.await(10, TimeUnit.SECONDS));
+
+        // Make sure there is no leftover junk left in the cluster state and task manager
+        assertCleanClusterState(0);
+        assertNoLeftOverTasks();
+    }
+
+    private static Map<String, Object> xContentToMap(ToXContent content) throws IOException {
+        XContentBuilder builder = XContentFactory.jsonBuilder();
+        if (randomBoolean()) {
+            builder.humanReadable(true);
+        }
+        builder.startObject();
+        content.toXContent(builder, EMPTY_PARAMS);
+        builder.endObject();
+        return XContentHelper.convertToMap(builder.bytes(), true).v2();
+    }
+
+    private long findTask(int node, Function<Task, Boolean> op) {
+        TaskManager taskManager = testNodes[node].transportService.getTaskManager();
+        for (Map.Entry<Long, Task> entry : taskManager.getTasks().entrySet()) {
+            if (op.apply(entry.getValue())) {
+                return entry.getKey();
+            }
+        }
+        return -1L;
+    }
+
+    private TransportTestCockroachAction[] registerTestCockroachAction(
+        BiFunction<TransportTestCockroachAction, TestRequest, TestResponse> nodeOperation,
+        BiFunction<TransportTestCockroachAction, TestResponse, TestResponse> callerOperation) throws InterruptedException {
+        TransportTestCockroachAction[] actions = new TransportTestCockroachAction[testNodes.length];
+        for (int i = 0; i < testNodes.length; i++) {
+            actions[i] = new TransportTestCockroachAction(Settings.EMPTY, threadPool, testNodes[i].transportService,
+                testNodes[i].clusterService, testNodes[i].cockroachService, testNodes[i].actionFilters,
+                testNodes[i].indexNameExpressionResolver, nodeOperation, callerOperation);
+        }
+        return actions;
+    }
+
+
+    private Task startTestCockroachAction(int coordinatorNode, TestRequest request,
+                                          BiFunction<TransportTestCockroachAction, TestRequest, TestResponse> nodeOperation,
+                                          BiFunction<TransportTestCockroachAction, TestResponse, TestResponse> callerOperation,
+                                          ActionListener<TestResponse> listener) throws InterruptedException {
+        TransportTestCockroachAction[] actions = registerTestCockroachAction(nodeOperation, callerOperation);
+        return actions[coordinatorNode].execute(request, listener);
+    }
+
+}
```

Comment:
I think we could also have tests for (unless I missed it): 
1. master changes while task executing (but before trying to send notification)
2. master changes while sending the response on the response node
3. master changes when task is marked as completed in cluster state but before the response is sent on the PublishingFailureListener